gh-134939: Add a Multiple Interpreters Howto Doc (#136143)
python/cpython: Doc/howto/multiple-interpreters.rst

.. TODO:

   * tutorial: explain how to use interpreters C-API?
   * add a top-level recipes/examples section
   * recipes: add some!
   * recipes: add short examples of how to solve specific small problems
   * recipes: add a section specifically for workflow-oriented examples
   * recipes: add comparisons with other concurrency models
   * recipes: add examples focused just on taking advantage of isolation?
   * recipes: add examples of using C-API in extensions?  embedded?

Introduction
============

You can find a thorough explanation about what interpreters are and
what they're for in the :ref:`concurrent.interpreters <interpreters-intro>`
docs.

In summary, think of an interpreter as the Python runtime's execution
context.  You can have multiple interpreters in a single process,
switching between them at any time.  Each interpreter is almost
completely isolated from the others.

.. note::

   Interpreters in the same process can technically never be strictly
   isolated from one another since there are few restrictions on memory
   access within the same process.  The Python runtime makes a best
   effort at isolation but extension modules may easily violate that.

>    Therefore, do not use multiple interpreters in security-senstive

Reviewer: ```suggestion
   Therefore, do not use multiple interpreters in security-sensitive
```

   situations, where they shouldn't have access to each other's data.

> That isolation facilitates a concurrency model based an independent

Reviewer: ```suggestion
That isolation facilitates a concurrency model based on independent
```

logical threads of execution, like CSP or the actor model.

Each actual thread in Python, even if you're only running in the main
thread, has its own *current* execution context.  Multiple threads can
use the same interpreter or different ones.

Why Use Multiple Interpreters?
------------------------------

These are the main benefits:

* isolation supports a human-friendly concurrency model
* isolation supports full multi-core parallelism
* avoids the data races that make threads so challenging normally

There are some downsides and temporary limitations:

* the concurrency model requires extra rigor; it enforces higher
  discipline about how the isolated components in your program interact
* not all PyPI extension modules support multiple interpreters yet
* the existing tools for passing data between interpreters safely

>   is still relatively inefficient and limited

Reviewer: ```suggestion
  are still relatively inefficient and limited
```

* actually *sharing* data safely is tricky (true for free-threading too)
* all necessary modules must be imported separately in each interpreter
* relatively slow startup time per interpreter
* non-trivial extra memory usage per interpreter

Nearly all of these should improve in subsequent Python releases.

.. _interp-tutorial-basics:

Tutorial: Basics
================

First of all, keep in mind that using multiple interpreters is like
using multiple processes.  They are isolated and independent from each
other.  The main difference is that multiple interpreters live in the

> same process, which makes it all more efficient and use fewer

Reviewer: ```suggestion
same process, which makes it all more efficient and uses fewer
```

    # prints: 0

Calling Methods and Other Objects in an Interpreter
---------------------------------------------------

Pretty much any callable object may be run in an interpreter, following
the same rules as functions::

    from concurrent import interpreters
    from mymod import Spam

    interp = interpreters.create()

    spam = Spam()
    interp.call(Spam.modname)
    # prints: mymod
    interp.call(spam)
    # prints: mymod
    interp.call(spam.spam)
    # prints: mymod

    class Eggs:
        @classmethod
        def modname(cls):
            print(__name__)
        def __call__(self):
            print(__name__)
        def eggs(self):
            print(__name__)

    eggs = Eggs()
    res = interp.call(Eggs.modname)
    # prints: <fake __main__>
    res = interp.call(eggs)
    # prints: <fake __main__>
    res = interp.call(eggs.eggs)
    # prints: <fake __main__>

    ##########
    # mymod.py
    ##########

    class Spam:
        @classmethod
        def modname(cls):
            print(__name__)
        def __call__(self):
            print(__name__)
        def spam(self):
            print(__name__)

Mutable State is not Shared
---------------------------

> Just be be clear, the underlying data of very few mutable objects is

Reviewer: ```suggestion
Just to be clear, the underlying data of very few mutable objects is
```

the other interpreter and never automatically synchronized::

    from concurrent import interpreters

    interp = interpreters.create()

    class Counter:
        def __init__(self, initial=0):
            self.value = initial
        def inc(self):
            self.value += 1
        def dec(self):
            self.value -= 1
        def show(self):
            print(self.value)
    counter = Counter(17)

    interp.call(counter.show)
    # prints: 17
    counter.show()
    # prints: 17

    interp.call(counter.inc)
    interp.call(counter.show)
    # prints: 18
    counter.show()
    # prints: 17

    interp.call(counter.inc)
    interp.call(counter.show)
    # prints: 18
    counter.show()
    # prints: 17

    interp.call(counter.inc)
    interp.call(counter.show)
    # prints: 18
    counter.show()
    # prints: 17

Preparing and Reusing an Interpreter
------------------------------------

When you use the Python REPL, it doesn't reset after each command.
That would make the REPL much less useful.  Likewise, when you use
the builtin :func:`exec`, it doesn't reset the namespace it uses
to run the code.

In the same way, running code in an interpreter does not reset that

> interpreter..  The next time you run code in that interpreter, the

Reviewer: ```suggestion
interpreter.  The next time you run code in that interpreter, the
```

thread the process started with.

In the same way, running code in another interpreter doesn't create
(or need) a new thread.  It uses the current OS thread, switching
to the interpreter long enough to run the code.

To actually run in a new thread, you can do it explicitly::

    from concurrent import interpreters
    import threading

    interp = interpreters.create()

    def run():
        interp.exec("print('spam!')")
    t = threading.Thread(target=run)
    t.start()
    t.join()

    def run():
        def script():
            print('spam!')
        interp.call(script)
    t = threading.Thread(target=run)
    t.start()
    t.join()

There's also a helper method for that::

    from concurrent import interpreters

    interp = interpreters.create()

    def script():
        print('spam!')
    t = interp.call_in_thread(script)
    t.join()

Handling Uncaught Exceptions
----------------------------

Consider what happens when you run a script at the commandline
and there's an unhandled exception.  In that case, Python will print
the traceback and the process will exit with a failure code.

The behavior is very similar when code is run in an interpreter.

> The traceback get printed and, rather than a failure code,

Reviewer: ```suggestion
The traceback gets printed and, rather than a failure code,
```

   situations, where they shouldn't have access to each other's data.

That isolation facilitates a concurrency model based on independent
logical threads of execution, like CSP or the actor model.

Each actual thread in Python, even if you're only running in the main
thread, has its own *current* execution context.  Multiple threads can
use the same interpreter or different ones.

Why Use Multiple Interpreters?
------------------------------

These are the main benefits:

* isolation supports a human-friendly concurrency model
* isolation supports full multi-core parallelism
* avoids the data races that make threads so challenging normally

There are some downsides and temporary limitations:

* the concurrency model requires extra rigor; it enforces higher
  discipline about how the isolated components in your program interact
* not all PyPI extension modules support multiple interpreters yet
* the existing tools for passing data between interpreters safely
  are still relatively inefficient and limited
* actually *sharing* data safely is tricky (true for free-threading too)
* all necessary modules must be imported separately in each interpreter

> * relatively slow startup time per interpreter

Reviewer: Some users might need a (foot)note that subinterpreters can sometimes improve the startup by importing nothing.

<details>
<summary>example</summary>

```py
def non_stateless():
    non_stateless

if __name__ == '__main__':
    import threading
    from concurrent import interpreters
    interps = [interpreters.create() for i in range(10)]
    threads = [
        threading.Thread(target=interp.call, args=(non_stateless,))
        for interp in interps
    ]
    for thread in threads:
        thread.start()
    for thread in threads:
        thread.join()
```
</details>

That seems not so beneficial for the `InterpreterPoolExecutor` case, though.

Author: I've added a note, though further down in the discussion about calling functions.

script in another interpreter.  We'll look more closely at that
case in a little while: :ref:`interp-script-args`.

Running in a Thread
-------------------

A key point is that, in a new Python process, the runtime doesn't
create a new thread just to run the target script.  It uses the main
thread the process started with.

In the same way, running code in another interpreter doesn't create
(or need) a new thread.  It uses the current OS thread, switching
to the interpreter long enough to run the code.

To actually run in a new thread, you can do it explicitly::

    from concurrent import interpreters
    import threading

    interp = interpreters.create()

    def run():
        interp.exec("print('spam!')")
    t = threading.Thread(target=run)
    t.start()
    t.join()

    def run():
        def script():
            print('spam!')
        interp.call(script)
    t = threading.Thread(target=run)
    t.start()
    t.join()

There's also a helper method for that::

    from concurrent import interpreters

    def script():
        print('spam!')

    if __name__ == '__main__':
        interp = interpreters.create()

>         t = interp.call_in_thread(script)

Reviewer: Off-topic, but would it be possible to add the `daemon` parameter to the helper method?

Author: I want to keep the helper super simple.  If someone wants a daemon thread then they can do it manually without much effort.

.. _multiple-interpreters-howto:

> ***************************
> Multiple Interpreters HOWTO
> ***************************

Reviewer 2: ```suggestion
***************************
Multiple interpreters HOWTO
***************************
```

In this HOWTO document we'll look at how to take advantage of
multiple interpreters in a Python program.  We will focus on doing
so in Python code, through the stdlib :mod:`concurrent.interpreters`
module.

> This document has 2 parts: first a brief introduction and then a tutorial.

Reviewer 2: ```suggestion
This document has two parts: first a brief introduction and then a tutorial.
```

Introduction
============

You can find a thorough explanation about what interpreters are and
what they're for in the :ref:`concurrent.interpreters <interpreters-intro>`
docs.

In summary, think of an interpreter as the Python runtime's execution
context.  You can have multiple interpreters in a single process,
switching between them at any time.  Each interpreter is almost
completely isolated from the others.

.. note::

   Interpreters in the same process can technically never be strictly
   isolated from one another since there are few restrictions on memory
   access within the same process.  The Python runtime makes a best
   effort at isolation but extension modules may easily violate that.
   Therefore, do not use multiple interpreters in security-sensitive
   situations, where they shouldn't have access to each other's data.

That isolation facilitates a concurrency model based on independent
logical threads of execution, like CSP or the actor model.

Each actual thread in Python, even if you're only running in the main
thread, has its own *current* execution context.  Multiple threads can
use the same interpreter or different ones.

> Why Use Multiple Interpreters?
> ------------------------------

Reviewer 2: ```suggestion
Why use multiple interpreters?
------------------------------
```

These are the main benefits:

* isolation supports a human-friendly concurrency model
* isolation supports full multi-core parallelism
* avoids the data races that make threads so challenging normally

There are some downsides and temporary limitations:

* the concurrency model requires extra rigor; it enforces higher
  discipline about how the isolated components in your program interact
* not all PyPI extension modules support multiple interpreters yet
* the existing tools for passing data between interpreters safely
  are still relatively inefficient and limited
* actually *sharing* data safely is tricky (true for free-threading too)
* all necessary modules must be imported separately in each interpreter
* relatively slow startup time per interpreter
* non-trivial extra memory usage per interpreter

Nearly all of these should improve in subsequent Python releases.

.. _interp-tutorial-basics:

> Tutorial: Basics

Reviewer 2: ```suggestion
Tutorial: basics
```

================

First of all, keep in mind that using multiple interpreters is like
using multiple processes.  They are isolated and independent from each
other.  The main difference is that multiple interpreters live in the
same process, which makes it all more efficient and uses fewer
system resources.

Each interpreter has its own :mod:`!__main__` module, its own
:data:`sys.modules`, and, in fact, its own set of *all* runtime state.
Interpreters pretty much never share objects and very little data.

> Running Code in an Interpreter
> ------------------------------

Reviewer 2: Use sentence case for headers:

https://devguide.python.org/documentation/style-guide/#capitalization

```suggestion
Running code in an interpreter
------------------------------
```

of those cases.

You can also exec any function that doesn't take any arguments, nor
returns anything.  Closures are not allowed but other nested functions
are.  It works the same as if you had passed the script corresponding
to the function's body::

    from concurrent import interpreters

    def script():
        print('spam!')

    def get_script():
        def nested():
            print('eggs!')
        return nested

    if __name__ == '__main__':
        interp = interpreters.create()

        interp.exec(script)
        # prints: spam!

        script2 = get_script()
        interp.exec(script2)
        # prints: eggs!

Any referenced globals are resolved relative to the interpreter's
:mod:`!__main__` module, just like happens for scripts, rather than
the original function's module::

    from concurrent import interpreters

    def script():
        print(__name__)

    if __name__ == '__main__':
        interp = interpreters.create()
        interp.exec(script)
        # prints: __main__

One practical difference is that with a script function you get syntax
highlighting in your editor.  With script text you probably don't.

.. _interp-tutorial-calls:

> Calling a Function in an Interpreter
> ------------------------------------

Reviewer 2: ```suggestion
Calling a function in an interpreter
------------------------------------
```

from there.  The exceptions are cases (3) and (4).

In case (3), the function is also copied, but efficiently and only
temporarily (long enough to be called) and is not actually bound to
any module.  This temporary function object's :data:`!__name__` will
be set to match the code object, but the :data:`!__module__`,
:data:`!__qualname__`, and other attributes are not guaranteed to
be set.  This shouldn't be a problem in practice, though, since
introspecting the currently running function is fairly rare.

We'll cover case (4) in
:ref:`the next section <interp-tutorial-funcs-in-main>`.

In all the cases, the function will get any global variables from the
module (in the target interpreter), like normal::

    from concurrent import interpreters
    with open('mymod.py', 'w') as outfile:
        outfile.write('def spam(): print(__name__)')
    from mymod import spam

    if __name__ == '__main__':
        interp = interpreters.create()
        interp.call(spam)
        # prints: mymod

    ##########
    # mymod.py
    ##########

    def spam():
        print(__name__)

Note, however, that in case (3) functions rarely look up any globals.

.. _interp-tutorial-funcs-in-main:

> Functions Defined in __main__

Reviewer 2: ```suggestion
Functions defined in __main__
```

-----------------------------

Functions defined in :mod:`!__main__` are treated specially by
:meth:`Interpreter.call`.  This is because the script or module that
ran in the main interpreter will have not run in other interpreters,
so any functions defined in the script would only be accessible by
running it in the target interpreter first.  That is essentially
how :meth:`Interpreter.call` handles it.

If the function isn't found in the target interpreter's :mod:`!__main__`
module then the script that ran in the main interpreter gets run in the

> target interpreter, though under a different name than "__main__".
> The function is then looked up on that resulting fake __main__ module.

Reviewer 2: ```suggestion
target interpreter, though under a different name than ":mod:`!__main__`".
The function is then looked up on that resulting fake :mod:`!__main__` module.
```

    total = 0

    def inc():
        global total
        total += 1

    def show_total():
        print(total)

    if __name__ == '__main__':
        interp = interpreters.create()

        interp.call(show_total)
        # prints: 0

        interp.call(inc)
        interp.call(show_total)
        # prints: 1

        interp.exec(dedent("""
            try:
                print(total)
            except NameError:
                pass
            else:
                raise AssertionError('expected NameError')
            """))
        interp.exec('total = -1')
        interp.exec('print(total)')
        # prints: -1

        interp.call(show_total)
        # prints: 1

        interp.call(inc)
        interp.call(show_total)
        # prints: 2

        interp.exec('print(total)')
        # prints: -1

        interp.exec('import sys; sys.stdout.flush()')

        print(total)
        # prints: 0

Note that the recommended ``if __name__ == '__main__`:`` idiom is
especially important for such functions, since the script will be
executed with :data:`!__name__` set to something other than
:mod:`!__main__`.  Thus, for any code you don't want run repeatedly,

> put it in the if block.  You will probably only want functions or

Reviewer 2: ```suggestion
put it in the ``if`` block.  You will probably only want functions or
```

classes outside the if block.

Another thing to keep in mind is that, when you run the REPL or run
``python -c ...``, the code that runs is essentially unrecoverable.  The
contents of the :mod:`__main__` module cannot be reproduced by executing
a script (or module) file.  Consequently, calling a function defined
in :mod:`__main__` in these cases will probably fail.

Here's one last tip before we move on.  You can avoid the extra
complexity involved with functions (and classes) defined in
:mod:`__main__` by simply not defining them in :mod:`__main__`.
Instead, put them in another module and import them in your script.

> Calling Methods and Other Objects in an Interpreter
> ---------------------------------------------------

Reviewer 2: ```suggestion
Calling methods and other objects in an interpreter
---------------------------------------------------
```

                    print(__name__)
                def spam(self):
                    print(__name__)
            """)
    from mymod import Spam

    class Eggs:
        @classmethod
        def modname(cls):
            print(__name__)
        def __call__(self):
            print(__name__)
        def eggs(self):
            print(__name__)

    if __name__ == '__main__':
        interp = interpreters.create()

        spam = Spam()
        interp.call(Spam.modname)
        # prints: mymod

        interp.call(spam)
        # prints: mymod

        interp.call(spam.spam)
        # prints: mymod

        eggs = Eggs()
        res = interp.call(Eggs.modname)
        # prints: <fake __main__>
        assert res is None

        res = interp.call(eggs)
        # prints: <fake __main__>
        assert res is None

        res = interp.call(eggs.eggs)
        # prints: <fake __main__>
        assert res is None

    ##########
    # mymod.py
    ##########

    class Spam:
        @classmethod
        def modname(cls):
            print(__name__)
        def __call__(self):
            print(__name__)
        def spam(self):
            print(__name__)

> Mutable State is not Shared
> ---------------------------

Reviewer 2: ```suggestion
Mutable state is not shared
---------------------------
```

        assert __name__ == '<fake __main__>', __name__
        global _counter
        _counter = counter

    def inc_global():
        assert __name__ == '<fake __main__>', __name__
        _counter.inc()

    def dec_global():
        assert __name__ == '<fake __main__>', __name__
        _counter.dec()

    def show_global():
        assert __name__ == '<fake __main__>', __name__
        _counter.show()

    if __name__ == '__main__':
        counter = Counter(17)
        interp = interpreters.create()

        interp.call(init_global, counter)

        interp.call(counter.show)
        # prints: 17
        interp.call(show_global)
        # prints: 17
        counter.show()
        # prints: 17

        interp.call(counter.inc)
        interp.call(counter.show)
        # prints: 17
        interp.call(inc_global)
        interp.call(show_global)
        # prints: 18
        counter.show()
        # prints: 17

        interp.call(counter.dec)
        interp.call(counter.show)
        # prints: 17
        interp.call(dec_global)
        interp.call(dec_global)
        interp.call(show_global)
        # prints: 16
        counter.show()
        # prints: 17

> Preparing and Reusing an Interpreter
> ------------------------------------

Reviewer 2: ```suggestion
Preparing and reusing an interpreter
------------------------------------
```

    if __name__ == '__main__':
        interp = interpreters.create()
        interp.call(set, 100)
        interp.call(check, 100)
        # prints: okay!

You can take advantage of this to prepare an interpreter ahead of time
or use it incrementally::

    from concurrent import interpreters
    from textwrap import dedent

    interp = interpreters.create()

    # Prepare the interpreter.
    interp.exec(dedent("""
        # We will need this later.
        import math

        # Initialize the value.
        value = 1

        def double(val):
            return val + val
        """))

    # Do the work.
    for _ in range(9):
        interp.exec(dedent("""
            assert math.factorial(value + 1) >= double(value)
            value = double(value)
            """))

    # Show the result.
    interp.exec('print(value)')
    # prints: 512

In case you're curious, in a little while we'll look at how to pass
data in and out of an interpreter (instead of just printing things).

Sometimes you might also have values that you want to use in a later
script in another interpreter.  We'll look more closely at that
case in a little while: :ref:`interp-script-args`.

> Running in a Thread
> -------------------

Reviewer 2: ```suggestion
Running in a thread
-------------------
```

A key point is that, in a new Python process, the runtime doesn't
create a new thread just to run the target script.  It uses the main
thread the process started with.

In the same way, running code in another interpreter doesn't create
(or need) a new thread.  It uses the current OS thread, switching
to the interpreter long enough to run the code.

To actually run in a new thread, you can do it explicitly::

    from concurrent import interpreters
    import threading

    interp = interpreters.create()

    def run():
        interp.exec("print('spam!')")
    t = threading.Thread(target=run)
    t.start()
    t.join()
    # prints: spam!

    def run():
        def script():
            print('spam!')
        interp.call(script)
    t = threading.Thread(target=run)
    t.start()
    t.join()
    # prints: spam!

There's also a helper method for that::

    from concurrent import interpreters

    def script():
        print('spam!')

    if __name__ == '__main__':
        interp = interpreters.create()
        t = interp.call_in_thread(script)
        t.join()
        # prints: spam!

> Handling Uncaught Exceptions
> ----------------------------

Reviewer 2: ```suggestion
Handling uncaught exceptions
----------------------------
```

> Consider what happens when you run a script at the commandline

Reviewer 2: ```suggestion
Consider what happens when you run a script at the command line
```

    interp.exec(dedent("""
        try:
            1/0
        except ZeroDivisionError:
            print('error ignored')
        """))
    # prints: error ignored

At the moment there isn't an easy way to "re-raise" an unhandled
exception from another interpreter.  Here's one approach that works for
exceptions that can be pickled::

    from concurrent import interpreters
    from textwrap import dedent

    interp = interpreters.create()

    try:
        try:
            interp.exec(dedent("""
                try:
                    1/0
                except Exception as exc:
                    import pickle
                    data = pickle.dumps(exc)
                    class PickledException(Exception):
                        pass
                    raise PickledException(data)
                """))
        except interpreters.ExecutionFailed as exc:
            if exc.excinfo.type.__name__ == 'PickledException':
                import pickle
                raise pickle.loads(eval(exc.excinfo.msg))
            else:
                raise  # re-raise
    except ZeroDivisionError:
        # Handle it!
        print('error ignored')
    # prints: error ignored

> Managing Interpreter Lifetime
> -----------------------------

Reviewer 2: ```suggestion
Managing interpreter lifetime
-----------------------------
```

Every interpreter uses resources, particularly memory, that should be
cleaned up as soon as you're done with the interpreter.  While you can
wait until the interpreter is cleaned up when the process exits, you
can explicitly clean it up sooner::

    from concurrent import interpreters

    interp = interpreters.create()
    try:
        interp.exec('print("spam!")')
    finally:
        interp.close()
    # prints: spam!

.. _interp-tutorial-communicating:

> Tutorial: Communicating Between Interpreters
> ============================================

Reviewer 2: ```suggestion
Tutorial: communicating between interpreters
============================================
```

Multiple interpreters are useful in the convenience and efficiency
they provide for isolated computing.  However, their main strength is
in running concurrent code.  Much like with any concurrency tool,
there has to be a simple and efficient way to communicate
between interpreters.

In this half of the tutorial, we explore various ways you can do so.

.. _interp-tutorial-return-values:

> Call Args and Return Values
> ---------------------------

Reviewer 2: ```suggestion
Call args and return values
---------------------------
```

between interpreters.  That means that, nearly always, arguments are
copied on the way in and return values on the way out.  Furthermore,
this will happen with each call, so it's a new copy every time
and no state persists between calls.

For example::

    from concurrent import interpreters

    data = {
        'a': 1,
        'b': 2,
        'c': 3,
    }

    if __name__ == '__main__':
        interp = interpreters.create()

        interp.call(data.clear)
        assert data == dict(a=1, b=2, c=3)

        def update_and_copy(data, **updates):
            data.update(updates)
            return dict(data)

        res = interp.call(update_and_copy, data)
        assert res == dict(a=1, b=2, c=3)
        assert res is not data
        assert data == dict(a=1, b=2, c=3)

        res = interp.call(update_and_copy, data, d=4, e=5)
        assert res == dict(a=1, b=2, c=3, d=4, e=5)
        assert data == dict(a=1, b=2, c=3)

        res = interp.call(update_and_copy, data)
        assert res == dict(a=1, b=2, c=3)
        assert res is not data
        assert data == dict(a=1, b=2, c=3)

        print('okay!')
        # prints: okay!

.. _interp-tutorial-shareable:

> Supported and Unsupported Objects
> ---------------------------------

Reviewer 2: ```suggestion
Supported and unsupported objects
---------------------------------
```

is a small set of other objects that are supported, such as
non-closure inner functions.

If you try to pass an unsupported object as an argument to
:meth:`Interpreter.call` then you will get a :exc:`NotShareableError`
with ``__cause__`` set appropriately.  :meth:`Interpreter.call` will
also raise that way if the return value is not supported.  The same
goes for :meth:`~Interpreter.prepare_main` and :class:`Queue.* <Queue>`.

Relatedly, there's an interesting side-effect to how we use a fake
module for objects with a class defined in :mod:`!__main__`.  If you
pass the class to :meth:`Interpreter.call`, which will create an
instance in the other interpreter, then that instance will fail to
unpickle in the original interpreter, due to the fake module name::

    from concurrent import interpreters

    class Spam:
        def __init__(self, x):
            self.x = x

    if __name__ == '__main__':
        interp = interpreters.create()
        try:
            spam = interp.call(Spam, 10)
        except interpreters.NotShareableError:
            print('okay!')
        else:
            raise AssertionError('unexpected success')
        # prints: okay!

> Sharing Data
> ------------

Reviewer 2: ```suggestion
Sharing data
------------
```

There are actually a small number of objects that aren't just copied and
for which the underlying data is actually shared between interpreters.
We'll talk about :class:`Queue` in the next section.

Another example is :class:`memoryview`, where the underlying
:ref:`buffer <bufferobjects>` can be read and written by multiple
interpreters at once.  Users are responsible to manage thread-safety
for that data in such cases.

Support for actually sharing data for other objects is possible through
extension modules, but we won't get into that here.

> Using Queues
> ------------

Reviewer 2: ```suggestion
Using queues
------------
```

:class:`concurrent.interpreters.Queue` is an implementation of the
:class:`queue.Queue` interface that supports safely and efficiently
passing data between interpreters::

    from concurrent import interpreters
    import time

    def push(queue, value, *extra, delay=0.1):
        queue.put(value)
        for val in extra:
            if delay > 0:
                time.sleep(delay)
            queue.put(val)

    def pop(queue, count=1, timeout=None):
        return tuple(queue.get(timeout=timeout)
                     for _ in range(count))

    if __name__ == '__main__':
        interp1 = interpreters.create()
        interp2 = interpreters.create()
        queue = interpreters.create_queue()

        t = interp1.call_in_thread(push, queue,
                                   'spam!', 42, 'eggs')
        res1 = interp2.call(pop, queue)
        res2 = interp2.call(pop, queue, 2)
        try:
            interp2.call(pop, queue, timeout=0)
        except interpreters.ExecutionFailed:
            pass
        t.join()
        print(res1)
        # prints: ('spam!',)
        print(res2)
        # prints: (42, 'eggs')

.. _interp-script-args:

> Initializing Globals for a Script
> ---------------------------------

Reviewer 2: ```suggestion
Initializing globals for a script
---------------------------------
```

    from concurrent import interpreters
    from textwrap import dedent

    interp = interpreters.create()
    queue = interpreters.create_queue()

    interp.prepare_main(queue=queue)
    queue.put('spam!')

    interp.exec(dedent("""
        msg = queue.get()
        print(msg)
        """))
    # prints: spam!

For basic, intrinsic data it can also make sense to use an f-string or
format string for the script and interpolate the required values::

    from concurrent import interpreters
    from textwrap import dedent

    interp = interpreters.create()

    def run_interp(interp, name, value):
        interp.exec(dedent(f"""
            {name} = {value!r}
            print('{name} =', {name})
            """))
    run_interp(interp, 'spam', 42)
    # prints: spam = 42

    with open('spam.txt', 'w') as outfile:
        outfile.write(dedent("""\
            line 1
            line 2
            EOF
            """))
    with open('spam.txt') as infile:
        interp.exec(dedent(f"""
            for line in open({infile.fileno()}, closefd=False):
                print(line, end='')
            """))
    # prints: line 1
    # prints: line 2
    # prints: EOF

> Pipes, Sockets, etc.
> --------------------

Reviewer 2: ```suggestion
Pipes, sockets, and so on
-------------------------
```

Avoid Latin abbreviations: https://devguide.python.org/documentation/style-guide/#use-simple-language

        r2, s2 = os.pipe()
        try:
            t = interp.call_in_thread(task, (r1, s2))

            # Initialize the worker.
            os.write(s1, INIT)
            os.read(r2, 1)

            steps = iter(steps)

            # Do the first step.
            for step in steps:
                # Do the step.
                print('...' if step is ... else step, end='')
                break
            for step in steps:
                # Let the worker take a step.
                os.write(s1, READY)
                os.read(r2, 1)

                # Do the step.
                print('...' if step is ... else step, end='')

            # Stop the worker.
            os.write(s1, STOP)

            # Wait for it to finish.
            os.read(r2, 1)
            t.join()
            print()
        finally:
            os.close(r1)
            os.close(s1)
            os.close(r2)
            os.close(s2)
    # prints: <a*b*c*...*x*y*z>

You can also close the pipe ends and join the thread to synchronize.

Using :func:`os.pipe` (or similar) to communicate between interpreters
is a little awkward, as well as inefficient.

.. _interp-tutorial-misc:

> Tutorial: Miscellaneous
> =======================

Reviewer 2: ```suggestion
Tutorial: miscellaneous
=======================
```
